lib/fetcher.js

```diff
@@ -203,7 +208,27 @@ class FetcherBase {
     // gets to the point of re-setting the integrity.
     const istream = ssri.integrityStream(this.opts)
     istream.on('integrity', i => this.integrity = i)
-    return stream.on('error', er => istream.emit('error', er)).pipe(istream)
+    // we have to return a stream that gets ALL the data, and proxies errors,
+    // but then pipe from the original tarball stream into the cache as well.
+    // To do this without losing any data, and since the cacache put stream
+    // is not a passthrough, we have to pipe from the original stream into
+    // the cache AFTER we pipe into the istream.
+    // If the cacache put stream was a passthrough, we could just return
+    // a single pipeline with all three, but that would make cacache put
+    // streams less useful, because they'd have to be read from to work.
+    stream.on('error', er => istream.emit('error', er))
+    stream.pipe(istream)
+    if (this.opts.cache && this[_cacheFetches]) {
+      const cstream = cacache.put.stream(
+        this.opts.cache,
+        `pacote:tarball:${this.from}`,
+        this.opts
+      )
+      // best-effort.  cache write errors should not crash the fetch.
+      cstream.on('error', (er) => {})
+      stream.pipe(cstream)
```

Comment:
I think the last piped stream needs to be returned here to have the caching actually run.

This fixes the problems I was seeing
```suggestion
      stream.pipe(cstream)
      return cstream
```

Alternatively:
```suggestion
      return stream.pipe(cstream)
```

Although you might want to just keep reassigning `stream` to the latest stream eg `stream = stream.pipe(istream)` & `stream = stream.pipe(cstream)`, `return stream`

Reply:
Returning `cstream` makes all the pacote tests fail, because the `cacache.put.stream` writable doesn't pass through its data.  Returning `cstream` is definitely not the right way to go.

If we pipe `istream` into `cstream` then all the data gets consumed before the caller can get access to it.  So as a result, you always get an empty tarball stream.

Reply:
Here's what I'm getting with this patch, using pacote directly via its bin:

```
$ node lib/bin.js tarball file:pacote-11.3.1.tgz copy.tgz --cache=$PWD/pacote-cache
{
  integrity: 'sha512-B5R+b6BX2U6exQ46KdYlHZNBsZL0RDe9TYwC6xbOgoU9JQ957m8zUxDAWwUuuCKb1JFPW+8KB+YCZRj4oBzzJQ==',
  resolved: '/Users/isaacs/dev/npm/pacote/pacote-11.3.1.tgz',
  from: 'file:pacote-11.3.1.tgz'
}

$ shasum -a 512 pacote-11.3.1.tgz
07947e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325  pacote-11.3.1.tgz

$ shasum -a 512 copy.tgz
07947e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325  copy.tgz

$ shasum -a 512 pacote-cache/content-v2/sha512/07/94/7e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325
07947e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325  pacote-cache/content-v2/sha512/07/94/7e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325

$ tree pacote-cache
pacote-cache
├── content-v2
│  └── sha512
│     └── 07
│        └── 94
│           └── 7e6fa057d94e9ec50e3a29d6251d9341b192f44437bd4d8c02eb16ce82853d250f79ee6f335310c05b052eb8229bd4914f5bef0a07e6026518f8a01cf325
├── index-v5
│  └── ee
│     └── 0d
│        └── 4c4d45587af9c1af774f9c59d3fc9b91b91764c81d2f3adb080948187863
└── tmp```

Reply:
Within npm/cli:

```$ npm i npm/pacote#pull/75

changed 1 package, and audited 907 packages in 3s

128 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ node . cache add file:npm-7.9.0.tgz --cache=test-cache

$ tree test-cache/
test-cache
└── _cacache
   ├── content-v2
   │  └── sha512
   │     └── c7
   │        └── ed
   │           └── 81d48a1de77d860a67af21f50f9dd65f1038ec8291b3b95a6eb665ad3c175d52a989fa10da3eddf2ed694c75dd53f4b6decdf2d43a9508982d846435246c
   ├── index-v5
   │  └── af
   │     └── 03
   │        └── 5c781820370e585dc2323edbbc80669bf714da5b47d56510c7d0bd7521ee
   └── tmp
      └── b1d942f0

$ shasum -a 512 test-cache/_cacache/content-v2/sha512/c7/ed/81d48a1de77d860a67af21f50f9dd65f1038ec8291b3b95a6eb665ad3c175d52a989fa10da3eddf2ed694c75dd53f4b6decdf2d43a9508982d846435246c
c7ed81d48a1de77d860a67af21f50f9dd65f1038ec8291b3b95a6eb665ad3c175d52a989fa10da3eddf2ed694c75dd53f4b6decdf2d43a9508982d846435246c  test-cache/_cacache/content-v2/sha512/c7/ed/81d48a1de77d860a67af21f50f9dd65f1038ec8291b3b95a6eb665ad3c175d52a989fa10da3eddf2ed694c75dd53f4b6decdf2d43a9508982d846435246c

$ shasum -a 512 npm-7.9.0.tgz
801b08dda0a4334e6643b666285d16f0e317c4d89bc0a5da1a0dd28306aa0b702f307b10fb270e230e364307bdb645ff16e88ac7611ef9702bf5b1bc926ef7b8  npm-7.9.0.tgz```

Seems like it works?

Reply:
Ah!  Spoke too soon.  The usage in npm/cli is breaking somehow.